### ISSUE TITLE
Edit and Continue

### ISSUE THREAD

**Issue by karthiknadig** [Microsoft/ptvsd#1192]
Allow code to be modified while at break-point, then continue executing the modified code.

**Comment by fabioz** [comment]
As a note, this is supported in the backend (reloading is not perfect, but good enough to be useful on many cases).

-- API: PyDevdAPI.request_reload_code(py_db, seq, module_name)

the main issue is that it doesn't seem like this there's a message from the debug adapter protocol notifying that a file was changed and should be reloaded, so, the client side would need to be implemented first and a custom message would be needed (i.e.: https://github.com/Microsoft/vscode/issues/6930#issuecomment-420798612)

**Comment by DonJayamanne** [comment]
The python extension can send these custom messages.

**Comment by int19h** [comment]
What are the limitations of the current implementation? I'm assuming that the code that is currently on the stack would not be affected, since there's no way to rewrite bytecode for a code object being executed, or adjust the state properly afterwards?

The key part here is establishing the expectations right. The most obvious way for a new user to try to do E&C is to hit a breakpoint in a function, and then try to edit the code *around the breakpoint*. But that would usually be in the same frame that's currently running.

Part of this is the feature name. The corresponding pre-existing feature for .NET languages in VS, while having many limitations, does allow this scenario, so long as you don't touch the currently executing *statement* (i.e. editing lines around it is fine). So if we call this E&C, we might be inadvertently setting up for expectations that are higher than what we can realistically deliver.

Perhaps "code reloading" would be a better name? This is more consistent with the terminology used in the Python ecosystem, and does not imply the ability to edit code that is currently actively executing.

**Comment by fabioz** [comment]
@int19h you're right, we don't rewrite the bytecode for the code object being executed (so, after doing the change the user needs to leave the function and enter it again -- usually that's good when you are in a webserver so you do a new request or have some UI application where you redo some action or if you step return and use a goto to go back to the previous line (maybe doing a goto to the end of the current function first) and enter the function again, so, these are the good use cases, but there are a many cases when that's not possible and I agree we should set the expectations properly here).

Also, the reload will patch function and classes in place, but there are limitations -- see: https://github.com/microsoft/ptvsd/blob/master/src/ptvsd/_vendored/pydevd/_pydevd_bundle/pydevd_reload.py#L70 for some limitations).

So, the good use case is when a function is changed and the user has a way to exit/enter it again, which I think is already pretty useful -- to be able to change what's being currently executed we'd need more support in CPython itself... maybe something to think about in the future ;)
